### PR TITLE
Search: Add ability to make authenticated API requests

### PR DESCRIPTION
**Pull request by nickdaugherty**
#### Changes proposed in this Pull Request:

Adds ability to make authenticated API requests from the JP site (via filtering jetpack_search_es_query_args) to support querying 'private' data (unpublished content for example)

#### Testing instructions:

* Add a filter on `jetpack_search_es_query_args` to set `authenticated_request` => `true`:

```
add-filter( 'jetpack_search_es_query_args', function( $args ) {
    $args['authenticated_request'] = true;

    return $args;
});
```

With the Search module active, perform some regular searches - requests should then go through `Jetpack_Client::wpcom_json_api_request_as_blog` class instead of `wp_remote_post()` and work normally.

### CODE REVIEW

File: modules/search/class.jetpack-search.php [hunk 1]
```diff
@@ -102,24 +102,49 @@ public function init_hooks() {
...
 	public function search( array $es_args ) {
-		$service_url = 'https://public-api.wordpress.com/rest/v1/sites/' . $this->jetpack_blog_id . '/search';
+		$endpoint    = sprintf( '/sites/%s/search', $this->jetpack_blog_id );
+		$service_url = 'https://public-api.wordpress.com/rest/v1' . $endpoint;
 
-		$start_time = microtime( true );
+		$do_authenticated_request = false;
+
+		if ( class_exists( 'Jetpack_Client' ) &&
+			isset( $es_args['authenticated_request'] ) &&
```
Comment (vindl) [inline modules/search/class.jetpack-search.php]: If we know that `$es_args['authenticated_request']` will assume boolean values, we could shorten this to:

`$do_authenticated_request = class_exists( 'Jetpack_Client' ) && ! empty( $es_args['authenticated_request'] );`

Comment (nickdaugherty) [reply modules/search/class.jetpack-search.php]: @vindl Thanks for looking! 

IMO it's better to be explicit and use strict equality, due to type juggling in PHP...since the string `"false"` evaluates to `true` and is not empty for example.

This way makes it a little harder to `_do_it_wrong()` as _only_ the expected value of `true` will cause the behavior. Since doing this could potentially expose otherwise non-public posts, it should be (relatively) hard to trigger. 

It's better if `$es_args['authenticated_requests'] = 'no';`, `$es_args['authenticated_requests'] = 'bananas';`, etc don't work :)

Comment (nickdaugherty) [reply modules/search/class.jetpack-search.php]: I also have to give credit to @mjangda who wrote the original code this was ported from.

Comment (vindl) [reply modules/search/class.jetpack-search.php]: It did occur to me while reviewing but I thought we can guarantee that that value will always be boolean. I agree with you though, better to be safe here than expose ourselves to that risk with the sole aim of saving a couple of lines.

**Comment by dereksmart** [comment]
Pushed to 5.3 7b8ccc84b
